api/capi/src/nnstreamer-capi-single.c

```diff
@@ -757,6 +839,10 @@ ml_single_close (ml_single_h single)
 
   /** locking ensures correctness with parallel calls on close */
   if (single_h->filter) {
+    g_list_foreach (single_h->destroy_data_list, __destroy_notify_free_data,
+        single_h);
+    g_list_free (single_h->destroy_data_list);
```

Comment:
Well.. I think this is unnecessary because you already freed all elements in `__destroy_notify_free_data`.

Or, if you want to keep this code style, `g_list_foreach (single_h->destroy_data_list, __destroy_notify, single_h)` is enough for your purpose, and remove unnecessary implementation of `__destroy_notify_free_data`

Reply:
I am not sure if I understand correctly.
`Glist` for each of its elements (`ml_tensors_data_s` here) allocates wrapper around them.
`g_list_foreach` above is supposed to free the `ml_tensors_data_s` structure.
`g_list_free` is supposed to free the wrapper made by the glist.
So, I am not on how to eliminate `g_list_free`?

You are right that `__destroy_notify_free_data` is a bit redundant. I removed this function, and reset `data->handle` in `__destroy_notify`. This allows calling `ml_tensors_data_destroy` with `g_list_free_full` which destroys `ml_tensors_data_s` wrapper along with glist wrapper.

Reply:
Oh, my bad. Yes, the double linked list in glib contains its wrapping object.